Make it explicit that tools can show the value of variables for asserts (#3641)
modelica/ModelicaSpecification: chapters/equations.tex

  \lstinline!condition! needs to be implicitly treated with \lstinline!noEvent! since otherwise events might be triggered that can lead to slightly changed simulation results.
  \end{nonnormative}
\end{itemize}

> Tools may show more than the message for failed assertions, in particular the condition and the values of any variables used in it.

Reviewer: I find this too weak for purposes of guiding how library authors should formulate assert messages.  Should they, or should they not explicitly include values of relevant variables in the message?  Should they even mention names of problematic model variables?

Author: I tried to extend that.

Author: Specifically I view it as two parts:
1. Models should just give a message. Nothing fancy - basically a literal string.
2. For users the information should be accessible, but I want to leave that open for tools to innovate.

Reviewer: These are good objectives, but I think it needs to be at least normatively recommended:
```suggestion
Tools are recommended to provide more information than just the given message of a failed assertion, in particular the condition and the values of variables used in it.
```

       AssertionLevel.warning);
assert(T > 200 and T < 500, "Medium model outside feasible region");
\end{lstlisting}

> It is recommended that models have a simple message as above and not include the value of any variables (like \lstinline!T!) in the message given to \lstinline!assert!.

Reviewer: Connect with normative recommendation above:
```suggestion
It is recommended that asserts have a simple message as above, formulated with the recommended tool behavior in mind.
```

> Tools are (especially for simple messages) recommended to automatically include the values of relevant variables, as part of the log or interactively accessible.
> Similarly tools are recommended to include the failed condition.

Reviewer: This is already part of the normative text above, so it needs to be removed or reformulated with the user in focus:
```suggestion
In this case, for example, embedding the value of \lstinline!T! in the message is likely to result in duplicated information and should be avoided.
```

Author: The idea with "especially for simple messages" was to allow tools to skip the duplicated T in such cases.

Author: Basically to allow good compatibility with existing models that already have such complicated messages, but provide a simple recommendation for users.

Author: Ok, tools may be smarter for a transition

Author: This has now been reformulated to be clearer (for tools and users) and include the point about duplicate information.

Reviewer: Resolving outdated conversation.

       AssertionLevel.warning);
assert(T > 200 and T < 500, "Medium model outside feasible region");
\end{lstlisting}

It is recommended that asserts have a simple message as above, formulated with the recommended tool behavior in mind.

> Tools are recommended to include the failed condition in the message.

Reviewer: This was already stated above.
```suggestion
```

> Tools are also (especially for literal strings) recommended to automatically include the values of relevant variables, as part of the log or interactively accessible.

Reviewer: Essentially already stated above.
```suggestion
```

> Writing \lstinline!assert(T<500, "Temperature = "+String(T)+" was above 500");! is thus not recommended, and is likely to lead to duplicated information -- although tools may have specific compatibility handling of such cases.

Reviewer: It should be sufficient to say _likely_; just let tools to their "clever" thing without mentioning it in the specification.  Also remove punctuation at end of inline code:
```suggestion
Writing \lstinline!assert(T<500, "Temperature = "+String(T)+" was above 500")! is thus not recommended, and is likely to lead to duplicated information.
```
